packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -490,197 +536,151 @@ export function getValidatorApi({
-        blindedBlock = null;
-        fullBlock = await fullBlockPromise;
-      } else {
-        throw Error(
-          `Internal Error: Neither builder nor execution proposal flow activated isBuilderEnabled=${isBuilderEnabled} builderSelection=${builderSelection}`
-        );
-      }
+        : Promise.reject(new Error("Builder disabled"));
 
-      const builderPayloadValue = blindedBlock?.executionPayloadValue ?? BigInt(0);
-      const enginePayloadValue = fullBlock?.executionPayloadValue ?? BigInt(0);
-      const consensusBlockValueBuilder = blindedBlock?.consensusBlockValue ?? BigInt(0);
-      const consensusBlockValueEngine = fullBlock?.consensusBlockValue ?? BigInt(0);
-
-      const blockValueBuilder = builderPayloadValue + consensusBlockValueBuilder; // Total block value is in wei
-      const blockValueEngine = enginePayloadValue + consensusBlockValueEngine; // Total block value is in wei
-
-      let executionPayloadSource: ProducedBlockSource | null = null;
-      const shouldOverrideBuilder = fullBlock?.shouldOverrideBuilder ?? false;
-
-      // handle the builder override case separately
-      if (shouldOverrideBuilder === true) {
-        executionPayloadSource = ProducedBlockSource.engine;
-        logger.info("Selected engine block as censorship suspected in builder blocks", {
-          // winston logger doesn't like bigint
-          enginePayloadValue: `${enginePayloadValue}`,
-          consensusBlockValueEngine: `${consensusBlockValueEngine}`,
-          blockValueEngine: `${blockValueEngine}`,
-          shouldOverrideBuilder,
-          slot,
-        });
-      } else if (fullBlock && blindedBlock) {
-        switch (builderSelection) {
-          case routes.validator.BuilderSelection.MaxProfit: {
+      const enginePromise = isEngineEnabled
+        ? produceEngineFullBlockOrContents(slot, randaoReveal, graffiti, {
+            feeRecipient,
+            strictFeeRecipientCheck,
+            // skip checking and recomputing head in these individual produce calls
+            skipHeadChecksAndUpdate: true,
+            commonBlockBody,
+          }).then((engineBlock) => {
+            // Once the engine returns a block, in the event of either:
+            // - suspected builder censorship
+            // - builder boost factor set to 0
+            // we don't need to wait for builder block as engine block will always be selected
             if (
-              // explicitly handle the two special values mentioned in spec for builder preferred / engine preferred
-              builderBoostFactor !== MAX_BUILDER_BOOST_FACTOR &&
-              (builderBoostFactor === BigInt(0) ||
-                blockValueEngine >= (blockValueBuilder * builderBoostFactor) / BigInt(100))
+              engineBlock.shouldOverrideBuilder ||
+              builderBoostFactor === BigInt(0) ||
+              builderSelection === routes.validator.BuilderSelection.ExecutionOnly
```

Comment:
We don't need to abort builder block if execution only as it does not even trigger both block production flows
```suggestion
              builderSelection === routes.validator.BuilderSelection.ExecutionAlways
```